### ISSUE TITLE
Topology templating

### ISSUE THREAD

**Issue by herlo** [CentOS-PaaS-SIG/linchpin#403]
Take the existing topologies, and allow for templating them in some way. The idea being that it would introduce a looping component and variable replacement similar to what Ansible does.

A simple use case for this might be to have three nodes in the same `resource_definitions`, but each of them be a different architecture. The architectures could either be passed in as an external variable (`{{ arches }}`) or it could be used in a loop at the bottom of the resource group.

```
---
libvirt:
  topology:
    topology_name: "libvirt"
    resource_groups:
      - resource_group_name: "ex"
        res_group_type: "libvirt"
        resource_definitions:
          - res_name: "centos71"
            type: "libvirt_node"
            uri: 'qemu:///system'
            count: 1
            image_src: 'http://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-1608.qcow2.xz'
            memory: 2048
            vcpus: 1
            arch: {{ item }}
            networks:
              - name: default
            additional_storage: 10G 
        with_items:
          - x86_64
          - ppc64le
          - s390x
```

The validation in this case might require that the items and the count match. This will take some more thinking and might not be the end result.

**Comment by herlo** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [[linchpin-dev] EPIC Dynamic Input and Templating](https://trello.com/c/nNfNmA05/253-linchpin-dev-epic-dynamic-input-and-templating)

**Comment by herlo** [comment]
Per conversations with @jaypoulz recently. Passing in variables is a useful thing. He's mostly interested in `count` and `arch` fields. 

Run linchpin would be something like:

`linchpin -vp PinFile.template --data 'arches=x86_64,s390x,ppc64le' up bkr-multiarch` 

The file `PinFile.template` would contain YAML like so:

```
---
bkr-multiarch:
  topology:
    topology_name: "bkr-march"
    resource_groups:
      - resource_group_name: "bkr-march"
        resource_group_type: beaker
        resource_definitions:
          - role: bkr_server
            whiteboard: Provisioned with linchpin
            job_group: ci-ops-central
            recipesets:
            {% for arch in {{ arches }} %}
              - distro: RHEL-6.5
                arch: {{ arch }}
                hostrequires:
                  - tag: processors
                    op: ">="
                    value: 1
            {% endfor %}
```

Would render to the following data, and passed to be acted upon:

```
bkr-multiarch:
  topology:
    topology_name: "bkr-march"
    resource_groups:
      - resource_group_name: "bkr-march"
        resource_group_type: beaker
        resource_definitions:
          - role: bkr_server
            whiteboard: Provisioned with linchpin
            job_group: ci-ops-central
            recipesets:
              - distro: RHEL-6.5
                arch: x86_64
                hostrequires:
                  - tag: processors
                    op: ">="
                    value: 1
              - distro: RHEL-6.5
                arch: ppc64le
                hostrequires:
                  - tag: processors
                    op: ">="
                    value: 1
              - distro: RHEL-6.5
                arch: s390x
                hostrequires:
                  - tag: processors
                    op: ">="
                    value: 1

```

The concept here is that the following template would be rendered

**Comment by herlo** [comment]
The flexibility of Jinja2 here allows for other options as well. Instead of wrapping the `recipesets` section as described above, a loop could be placed around all `resource definitions`, here's a libvirt example:

Additional options for the command line would be something like this:

`linchpin -vp PinFile.template --data /path/to/datafile.yml up libvirt-mi`

The content of /path/to/datafile.yml would be:

```
---
resources:
  - name: centos71
    src: http://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-1608.qcow2.xz
  - name: centos65
    src: http://cloud.centos.org/centos/6.5/images/CentOS-6-x86_64-GenericCloud-20140929_01.qcow2
  - name: centos66
    src: http://cloud.centos.org/centos/6.6/images/CentOS-6-x86_64-GenericCloud-1711.qcow2.xz
  - name: fedora27
    src: https://download.fedoraproject.org/pub/fedora-secondary/releases/27/Cloud/aarch64/iso/Fedora-Cloud-netinst-aarch64-27-1.6.iso
    arch: aarch64

```
And the Pinfile.template file would be:

```
---
libvirt-mi:
  topology:
    topology_name: "libvirt-multi"
    resource_groups:
      - resource_group_name: "libvirt-mi"
        resource_group_type: "libvirt"
        resource_definitions:
        {% for resources in resources %}
          - role: libvirt_node
            name: {{ resource.name }}
            image_src: {{ resource.src }}
            memory: 1024
            vcpus: 1
            arch: {{ resource.arch | default('x86_64') }}
            networks:
              - name: default
        {% endfor %}
```

The processed data passed to linchpin up would be:

```
libvirt-mi:
  topology:
    topology_name: "libvirt-multi"
    resource_groups:
      - resource_group_name: "libvirt-mi"
        resource_group_type: "libvirt"
        resource_definitions:
          - role: libvirt_node
            name: centos71
            image_src: http://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-1608.qcow2.xz
            memory: 1024
            vcpus: 1
            arch: x86_64
            networks:
              - name: default
          - role: libvirt_node
            name: centos65
            image_src: http://cloud.centos.org/centos/6.5/images/CentOS-6-x86_64-GenericCloud-20140929_01.qcow2
            memory: 1024
            vcpus: 1
            arch: x86_64
            networks:
              - name: default
          - role: libvirt_node
            name: centos66
            image_src: http://cloud.centos.org/centos/6.6/images/CentOS-6-x86_64-GenericCloud-1711.qcow2.xz
            memory: 1024
            vcpus: 1
            arch: x86_64
            networks:
              - name: default
          - role: libvirt_node
            name: fedora27
            image_src: https://download.fedoraproject.org/pub/fedora-secondary/releases/27/Cloud/aarch64/iso/Fedora-Cloud-netinst-aarch64-27-1.6.iso
            memory: 1024
            vcpus: 1
            arch: aarch64
            networks:
              - name: default
```

**Comment by herlo** [comment]
Closed by #425
